### ISSUE TITLE
Use FilterQuery Type to modify document property throw Error.

### ISSUE THREAD

**Issue by yusunglee2074** [Automattic/mongoose#10186]
**Do you want to request a *feature* or report a *bug*?**
Bug
**What is the current behavior?**

Use FilterQuery Type to modify document property. 

```
import { Document, Schema as MongooseSchema } from 'mongoose';

export type UserDocument = User & Document;

export class User {
  email: string;

  password: string;
}

async update(
  filter: FilterQuery<UserDocument>,
  data: UpdateQuery<UserDocument>,
) {
  if (data.password) {
    let password: string;
    data.password = hashPassword(data.passward)
  }
  this.userModel
      .findOneAndUpdate(filter, data.password ? { ...data, password } : data, {
        new: true,
      })
}
```

Errors.
Property 'password' does not exist on type 'UpdateQuery<UserDocument>
Property 'password' does not exist on type 'UpdateWithAggregationPipeline'.

tsconfig.json
```
{
  "compilerOptions": {
    "module": "commonjs",
    "declaration": true,
    "removeComments": true,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "allowSyntheticDefaultImports": true,
    "target": "es2017",
    "sourceMap": true,
    "outDir": "./dist",
    "baseUrl": "./",
    "incremental": true
  }
}
```

**What is the expected behavior?**
I can modify exist UserDocument property.

Node.js: v14.16.0
mongoose: v5.12.5 and v5.12.6 (After 5.12.5 error occurs, after this commit) https://github.com/Automattic/mongoose/commit/61d24d41e00454fccda9bf1e69730d2bdbe700c9#commitcomment-50023157

**Comment by yusunglee2074** [comment]
```
error TS2339: Property 'password' does not exist on type 'UpdateQuery<UserDocument>'.
  Property 'password' does not exist on type 'UpdateWithAggregationPipeline'.
65     if (data.password) {
```

**Comment by vkarpov15** [comment]
I can confirm the below script works once c4ba0e9 is applied:

```javascript
import { Document, Schema as MongooseSchema, FilterQuery, UpdateQuery } from 'mongoose';
  
export type UserDocument = User & Document;

export class User {
  email: string;

  password: string;
}

async function update(
  filter: FilterQuery<UserDocument>,
  data: UpdateQuery<UserDocument>,
) {
  let password = '42';
  this.userModel
      .findOneAndUpdate(filter, data.password ? { ...data, password } : data, {
        new: true,
      })
}
```

Fix will be in v5.12.8.
